pallets/stateful-storage/src/tests.rs

```diff
@@ -0,0 +1,2533 @@
+	new_test_ext().execute_with(|| {
+		// arrange
+		let caller_1 = test_public(1);
+		let (msa_id, pair) = get_signature_account();
+		let delegator_key = pair.public();
+		let schema_id = UNDELEGATED_PAGINATED_SCHEMA;
+		let page_id = 1;
+		let block_number = 10;
+		let payload = generate_payload_bytes::<PaginatedPageSize>(Some(100));
+		let payload = PaginatedUpsertSignaturePayload {
+			payload,
+			target_hash: PageHash::default(),
+			msa_id,
+			expiration: block_number,
+			schema_id,
+			page_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = pair.sign(&encode_data_new_key_data).into();
+
+		// act
+		System::set_block_number(block_number);
+		assert_err!(
+			StatefulStoragePallet::upsert_page_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::ProofHasExpired
+		)
+	})
+}
+
+#[test]
+fn upsert_page_with_signature_having_out_of_window_payload_should_fail() {
+	new_test_ext().execute_with(|| {
+		// arrange
+		let caller_1 = test_public(1);
+		let (msa_id, pair) = get_signature_account();
+		let delegator_key = pair.public();
+		let schema_id = UNDELEGATED_PAGINATED_SCHEMA;
+		let page_id = 1;
+		let payload = generate_payload_bytes::<PaginatedPageSize>(Some(100));
+		let payload = PaginatedUpsertSignaturePayload {
+			payload,
+			target_hash: PageHash::default(),
+			msa_id,
+			expiration: (<Test as Config>::MortalityWindowSize::get() + 1).into(),
+			schema_id,
+			page_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = pair.sign(&encode_data_new_key_data).into();
+
+		// act
+		assert_err!(
+			StatefulStoragePallet::upsert_page_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::ProofNotYetValid
+		)
+	})
+}
+
+#[test]
+fn upsert_page_with_signature_having_wrong_signature_should_fail() {
+	new_test_ext().execute_with(|| {
+		// arrange
+		let caller_1 = test_public(1);
+		let (msa_id, pair) = get_signature_account();
+		let delegator_key = pair.public();
+		let (signature_key, _) = sr25519::Pair::generate();
+		let schema_id = UNDELEGATED_PAGINATED_SCHEMA;
+		let page_id = 1;
+		let payload = generate_payload_bytes::<PaginatedPageSize>(Some(100));
+		let payload = PaginatedUpsertSignaturePayload {
+			payload,
+			target_hash: PageHash::default(),
+			msa_id,
+			expiration: 10,
+			schema_id,
+			page_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = signature_key.sign(&encode_data_new_key_data).into();
+
+		// act
+		assert_err!(
+			StatefulStoragePallet::upsert_page_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::InvalidSignature
```

Comment:
Is this situation possible?

1. Signature 0 to approve change to `state Null -> A`
2. Signature 1 to approve change to `state A -> B`
3. Signature 2 to approve change to back to A: `state B -> A`
4. Reuse signature 1 (`state A->B`) to force the user to move back again to `state B`

Reply:
That's why we need the ringbuffer. I think we need the same replay prevention logic that's in the msa pallet. Or, probably, given the way things will be structured with Capacity, maybe we need to refactor the signature replay so it can be easily shared.

Reply:
This is a possibility if two signatures 1 and 2 are signed close to each other. We can reduce the possibility by choosing a narrower signature validation window. 
I think this should be a rare case since you would need all of following conditions to happen:

1. two signatures signed inside defined mortality window.
2. second signature should happen after first signature since it should know the hash. (or can predict it)
3. second signature should reverse the state to exactly a previous version inside the mortality window.

I think due to the nature of data in graph we can get away with it. This might be more important for itemized keys but since we are applying append only for those it can not go back to a previous state.

Anyway I don't think adding additional DB reads and writes to handle this is feasible.

Reply:
I agree that this doesn't need as much of a protection as account control does like in the msa pallet. Also anything that would be problematic to reuse a signature for should use the Append Only permission.

I think the options are:
1. Leave as is. Some schemas might not have this concern or can use append only if they do.
2. Skip the requirement of ability to calculate the state hash off-chain and add the last edited block number to the data. (once per page and once for the entire itemized list). (Allows add page->delete page->add page replay still, but removes all others)
3. Almost the same as 2, but use a nonce instead of the hash and don't use a hash at all. Just use the nonce. (Allows add page->delete page->add page replay still, but removes all others)
4. (I agree with Aramik, not an option from performance requirements) Use a separate piece of state to handle it.
5. Require signature would also require append only.
6. (Already rejected. Causes any state change to be dependent on others and doesn't work with the user-provider model) Add a nonce to the msa or such.

I wonder if 3 gives us any other improvements in that we don't have to hash the data, but that's a tiny amount of savings.

I think I am currently leaning toward 1, but maybe 5 or 3? Curious if there are other ideas.

Reply:
there was a conversation with @wesbiggs about using block number or hash and hash was superior based on some scenarios. One of them was the following:

> Wes Biggs:
> The only problem I see with block number from the DSNP perspective is that an application that is making a series of 
> writes over the same session (but in different blocks) will not know what block its previous change was part of (without always querying for state before submitting any updates). 
> OK yeah, we might be talking about different problems. Here's my scenario:
> Alice uses DSNP through two different providers' apps. Both apps allow her to manipulate her social graph, and she uses them interchangeably to do so. We want to ensure that if the cached graph in either app gets stale, this condition will be detected at write time and not allow changes made in App 1 to overwrite changes that were made, unbeknownst to App 1, in App 2.
> 1. [Block 0] Suppose Alice's starting graph contains { Bob, Carol }
> 2. [Block 1] Alice opens App 1. Provider 1 fetches Alice's current graph, and caches { Bob, Carol }
> 3. [Block 2] Alice opens App 2. Provider 2 fetches Alice's current graph, and caches { Bob, Carol }
> 4. [Block 3] Alice adds a connection to Dan in App 2. App 2's cache now has { Bob, Carol, Dan }. Provider 2 writes { Bob, Carol, Dan } to the chain.
> 5. [Block 4] Alice switches back to App 1. She adds a connection to Erin. App 1's cache now has { Bob, Carol, Erin }. App 2 attempts to write { Bob, Carol, Erin } to the chain.

> We want the chain to reject 5 because its preimage, { Bob, Carol } is out of sync.

Reply:
> Require signature would also require append only.

Having append only for paginated schemas it a bit weird. For current use cases it seems fine since we need both of them for adding keys but I can see that in future we might be able to use signature-required ones even for paginated for cost shifting or any other reason.

Reply:
@wilwade I agree--the current interpretation of `AppendOnly` for paginated is really `UpsertOnly`. Which doesn't really seem to provide a benefit (we can't delete a page, but we can overwrite it as empty...). Maybe we should enforce in the `schemas` pallet that `AppendOnly` can only be applied to `PayloadLocation::Itemized` and update the attribute documentation accordingly.

Reply:
Decided to spike using a nonce inside the page so that the overall state can not back to a previous one without page removal.

Reply:
Fixed via https://github.com/LibertyDSNP/frequency/pull/1220